src/utils/timer.ts

```diff
@@ -0,0 +1,105 @@
+import {
+  addTimedelta,
+  TimeComponents,
+  toMilliseconds
+} from "./time"
+
+export type TimerComponentsParameters = {
+  duration: TimeComponents,
+  tick?: TimeComponents,
+  continues?: boolean,
+  onTick?: (timer: Timer) => void,
+  onComplete: (timer: Timer) => void,
+  onStart?: (timer: Timer) => void,
+}
+
+/**
+ * Generic timer implementation
+ * @author Mike Reiche <mike@reiche.world>
+ */
+export class Timer {
+  private readonly _params: TimerComponentsParameters
+  private _tickInterval: number
```

Comment:
Might be nice to use a little reverse hungarian here. Reading the name, I thought this would be a synthetic clock resolution. But it's actually the handle to the browser-internal timer. (this._params.tick is the clock resolution.)

`tickIntervalHandle` (or maybe just `intervalHandle`, given we only have one, or maybe `tickHandle`, given that "interval" is implicit in handling ticks).

Side-note: Underscore is interpreted by TypeScript to mean "ignore unused", so it's not a good naming convention (in TypeScript) for private variables. (It looks like this will [not change in the near future](https://github.com/microsoft/TypeScript/issues/29202).)
If they are private, there is never a reason to keep the unused ones. TypeScript _does_ [support `#` as a prefix](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields) for private variables, but personally, I'd just [declare them as `private`](https://www.typescriptlang.org/docs/handbook/2/classes.html#private) and be done with it.